\documentclass[twocolumn,10pt]{scrartcl}

% packages
\usepackage{xltxtra}
\usepackage{polyglossia}
\usepackage[left=2cm,right=2cm,top=3cm,bottom=3cm]{geometry}
\usepackage{hyperref}
\usepackage{natbib}
% fonts general
%\setmainfont[Mapping=tex-text,Scale=1.0]{FreeSerif}
%\setsansfont[Mapping=tex-text,Scale=1.0]{FreeSans}
%\setmonofont{FreeMono}


\begin{document}

\title{BEASTling: a software tool for linguistic phylogenetics using BEAST 2}
\author{Luke Maurits, Robert Forkel, Gereon Kaiping, Quentin Atkinson, Russell Gray}
\maketitle

\abstract{\textbf{Abstract:} We present a new open source software tool called BEASTling, designed to simplify the preparation of Bayesian phylogenetic analyses of linguistic data using the BEAST 2 platform.  BEASTling transforms comparatively short and human-readable configuration files into the XML files used by BEAST to specify analyses.  By taking advantage of Creative Commons-licensed data from the Glottolog language catalog, BEASTling
  % facilitates automatic and intelligent handling of datasets,
  allows the user to conveniently filter datasets using names for recognised language families, to impose monophyly constraints so that inferred language trees are backward compatible with Glottolog classifications, or to assign geographic location data to languages for phylogeographic analyses.
  Support for the emerging cross-linguistic linked data format (CLDF) permits easy incorporation of data published in cross-linguistic linked databases into analyses.
  BEASTling is intended to make the power of Bayesian analysis more accessible to historical linguists without strong programming backgrounds, in the hopes of encouraging communication and collaboration between those developing computational models of language evolution (who are typically not linguists) and relevant domain experts.}
\section{Introduction}

Recent years have seen an increased interest in the use of computational and especially Bayesian methods for inferring phylogenetic trees of languages within an explicit, model-based framework \citep[e.g.,][]{Gray2003,Gray2009,Kitchen2009,Dunn2011a,Lee2011,Walker2011,Bouckaert2012,Bowern2012,Currie2013,Honkola2013,Sicoli2014,Grollemund2015,Chang2015}.  Given the recency of this trend, there is currently a lack of linguistics-specific software for these sorts of analyses.  Thus, published analyses to date have all relied on software developed for biological phylogenetics, such as BayesTraits\cite{Pagel2004}, BEAST\cite{Drummond2012,Bouckaert2014} or MrBayes\cite{Huelsenbeck2001, Ronquist2003}.

Amongst these existing pieces of software, BEAST (in its modern 2.x form) is unique in that it was deliberately designed from the ground up to support user extensibility.  It is possible for users to write packages which extend the phylogenetic modelling capabilities of BEAST, by e.g. specifying new tree priors or new substitution models.  This makes BEAST an appealing platform for the burgeoning field of computational linguistic phylogenetics, as those working in the field have the ability to develop language-specific models and readily share them amongst themselves, in a process requiring no input from the BEAST development team.

One impediment to the widespread adoption of BEAST by the linguistics community is the difficulty of configuring analyses.  BEAST analyses are specified in Extensible Markup Language (XML) files which contain the input data as well as all the details of the modelling, including things such as tree constraints, substitution models, clock and mutation rate variation, prior distributions for trees and all parameters, as well as the various mechanisms for drawing MCMC proposals, details on which parameters to log, etc.

The recommended way for BEAST users to generate these XML analysis specifications is using the BEAUti program distributed with BEAST.  BEAUti is a GUI tool intended to provide a user friendly means of generating XML files for BEAST.  However, BEAUti is not an especially appealing tool for linguists.  Because BEAST originates as a tool for biological phylogenetics, BEAUti makes extensive use of biological terminology which may be unfamiliar or misleading to linguists; in particular, words like \emph{alignment} or \emph{morphology} have different meanings in the two fields.  BEAUti expects to load data in the NEXUS file format\cite{Maddison1997}, which is commonly used in biology but not in linguistics.  The workflow in BEAUti is best suited to scenarios where the same modelling assumptions apply to a very large number of datapoints of the same kind, such as individual nucleotides in a DNA or RNA sequence.  This is very different to one typical situation in linguistics, where each feature in a typological dataset may warrant its own unique treatment, which the BEAUti workflow is not well suited to.  Further, many desirable tasks for setting up a linguistic analysis are tedious and error-prone procedures in BEAUti.  For example, placing monophyly constraints so that all of the languages in a well supported and recognised language family are forced to remain in a single clade requires using the mouse to find and individually click on the name of every language in the family, choosing them out of a scrolling list of all languages in the analysis.  The same task is required for adding calibration dates to the ancestor of sets of families.  This is impractical for large and detailed analyses.

Among experienced BEAST users in the computational biology field, it is a common practice to write small scripts which parse data files and generate XML files for complex analyses which are impractical or impossible to configure using BEAUti.  The required degree of programming ability, familiarity with XML and knowledge of BEAST internals present a high barrier of entry to interested linguists who might otherwise bring valuable experience to the field.

In addition to BEAST XML files being difficult to generate, they are also not very human-readable.  Many of the XML elements have names or attributes which are derived from the names of objects, methods or parameters in the BEAST source code, which most users are not familiar with.  The XML files for complicated analyses of non-homogenous datasets, as frequently encountered in linguistics, can easily be thousands of lines long.  Reading an unfamiliar XML file - such as one provided by a collaborator, one a user generated themselves a year ago for a old project, or one associated with a manuscript undergoing peer review - and understanding all of the details of the analysis is a difficult and time-consuming process, especially for inexperienced users.  Conceptually simple changes to an analysis, such as removing a single language or linguistic feature, can require deleting or carefully changing multiple lines of the XML file, spread over several different locations.  While publishing a BEAST XML file in principle provides a level of methodological transparency and reproducability which is commendable and rarely achieved in scientific research, the practical difficulty of working with the files might sometimes discourage close scrutiny, error checking, or making small changes to investigate the effect of different assumptions.

In this paper we introduce a new software package named \emph{BEASTling} which aims to alleviate these problems.  The goal of BEASTling is to make the specification of BEAST analyses for linguistic phylogenetics simple and accessible, even when the analyses are large or complicated.  By making the power of BEAST accessible to a broader range of interested researchers in linguistics, we hope to foster increased collaboration between experts in linguistics and computational evolution to further and faster develop a formal, quantitative and data-driven approach to historical linguistics.

\section{Design and Implementation}

BEASTling is a command-line program written in the Python programming language\footnote{Python versions 2.7 and 3.4+ are supported.}, which takes the approach of transforming one kind of configuration file into another, effectively specifying an alternative configuration file format for BEAST.  In contrast to BEAST's native XML files, BEASTling configuration files are designed to be short, neat, human-readable, and intuitively editable by hand.  Analyses are specified at a high-level of abstraction, roughly corresponding to the details that a typical user might have in their head when designing an analysis, e.g. ``I'd like to infer a phylogenetic tree of the Austronesian languages, using structural data from WALS and the Lewis Mk substitution model, with a relaxed clock and rate variation across features''.  Typical BEASTling configuration files are approximately 15-25 lines long.  An example valid configuration file, implementing the Austronesian analysis described above, is shown in Figure \ref{fig:config}.

\begin{figure}[t]
	\begin{verbatim}[admin]
basename = austronesian_wals
[MCMC]
chainlength = 50000000
[languages]
families = Austronesian
monophyly = True
[model wals]
data = wals_data.csv
model = mk
frequencies = empirical
rate_variation = True
[clock default]
type = relaxed
[calibration]
root = 4700 - 5700\end{verbatim}
	\caption{An example BEASTling configuration file, specifying an analysis of the Austronesian language family using structural data from WALS, with internal structure partially constrained by Glottolog. A relaxed clock model is used, and the data is modelled by the Lewis Mk model with empirical base frequencies and Gamma-distributed rate variation across features.  A calibration date range is applied to the root of Austronesian.}
\label{fig:config}
\end{figure}

When a BEASTling configuration file is transformed into a BEAST XML file, the text of the original BEASTling configuration is embedded at the start of the XML file in a comment block, along with the date and time the transformation was performed and the version of BEASTling which was used.  This goes a long way to making the generated XML files self-documenting.  An old, forgotten analysis file can be understood in one minute instead of one hour, and making tweaks is straightforward as the old configuration can be copied into a new file and modified.  Furthermore, BEASTling provides an option to embed copies of all referenced data files in the XML file as additional comment blocks.  The resulting XML file is then entirely self-contained, allowing researchers to distribute in a single file everything required to precisely reproduce and to modify the analysis in question.  In addition to the XML file, BEASTling can optionally also produce an ``executive summary'' report (in MarkDown format) detailing the languages and modelling involved in an analysis, and a map of the languages' locations (in GeoJSON format).  Both of these files will be automatically rendered by GitHub if committed to a repository there, encouraging users to share their analyses in this way, which in turn encourages replication and modification of published analyses.

BEASTling analyses are focussed on the inference of phylogenetic trees from linguistic data.  Sampled trees are logged by default, in the standard Newick format.  Additional information can be logged if desired, including things like clock and substitution model parameters, and the ages of certain clades.  It is also possible for users to provide a known and trusted phylogenetic tree which is held fixed during the analysis, so that model parameters may be estimated conditioned on that tree.  Noteworthy modelling features supported by the current BEASTling release (1.2.0) are Gamma-distributed rate heterogeneity across features, relaxed and random local clocks, age calibrations on clades, monophyly constraints, a range of substitution models and one phylogeographic model, which supports spatial calibrations on clades.  In addition, BEASTling can automatically handle many data-preprocessing chores such as automatically filtering a dataset down to only a specified set of languages and/or features, removing languages which have only missing data for the selected feature set, removing features which have only missing data or constant values for the selected language set, and recoding cognate judgements into independent binary features.

\subsection{Phylogenetic modelling}

BEASTling makes what are intended to be sensible high-level decisions about the nature of the probabilistic phylogenetic models underlying its analyses, while still providing sufficient control and flexibility that appropriate models can be specified for most investigations.

BEASTling analyses use the Yule pure birth process\cite{Yule1924} to define a prior distribution over phylogenetic trees.  The birthrate parameter is constant over all locations on the tree, but the particular constant value is inferred during the MCMC procedure.  The Yule prior is one of two tree prior families supported by BEAST, and in biological applications is typically used to constrain trees over multiple species, i.e. the branching events are interpreted as speciation.  The other supported family is the coalescent process\cite{Kingman1982}, which is typically used for trees over populations of a single species, i.e. the branching evens are interpreted as reproduction.  Coalescent trees have a characteristic shape in which the oldest branching events are very much older than the most recent.  There is no theoretical basis for expecting the language diversification process, which is more often analogised to speciation than within-population variation, to yield trees with this shape, nor is there empirical evidence in any established reconstructions.  BEASTling therefore prefers the Yule prior.  One shortcoming of this approach is that the Yule model assumes that languages never go extinct, when in fact language extinction is believed to be a frequent occurrence.  The development of new tree priors specifically designed for linguistic phylogenetics is a continuing area of research, and future releases of BEASTling will include support for any suitable new tree priors implemented for BEAST.
% Cite Yule tree, coalescent tree and a historical reconstruction that is more Yule-like than Coalescent-like.

BEASTling supports a number of different clock models, for controlling how tree branch lengths are converted into a measure of evolutionary time; strict clocks, where the same constant rate is applied all over the tree, relaxed clocks\cite{Drummond2006}, where each branch has its own rate sampled from a tree-wide distribution (Lognormal, Exponential and Gamma distributions are supported), and random local clocks\cite{Drummond2010} which interpolate between these two possibilities.  If no calibration dates are provided, then the rate of a strict clock or the mean rate of a relaxed or random local clock is fixed at 1.0, and the branch lengths of the tree can then be interpreted as having units of ``expected number of substitutions per feature''.  However, if calibration dates are provided for any clades in the tree, then the tree branch lengths are in the same units as the calibration dates, and the appropriate corresponding clock rate is inferred.  Date calibrations can be combined with Yule priors exactly in certain cases\cite{Heled2011} and only approximately in others.  Both approaches are supported by BEAST, and BEASTling automatically applies the exact method when appropriate.

In addition to variation in clock rate over the tree, BEASTling also provides support for variation in substitution rate across different linguistic features.  If enabled, each feature is assigned its own substitution rate whose value is inferred during the analysis.  Substitution rates are assigned a Gamma-distributed prior with a mean value of 1.0.  The mean of all substitution rates is also constrained to be 1.0, so that the resulting rates are easily interpreted as rates relative to the average rate, e.g. a feature with rate 2.4 evolves at more than twice the average speed, while a rate of 0.13 is almost ten times slower than average.  The Gamma distribution's shape parameter, which determines the amount of variation in rate, is fitted to the data.  The need to support variation in substitution rate across features is very well established in linguistics.  The assumption that no such variation exists was one of the strongest and earliest criticisms of the first formulations of glottochronology, and modern analyses have demonstrated the existence of considerable rate variation for both lexical\cite{Pagel2007} and typological\cite{Greenhill2010,Dediu2011} data.

The probabilistic models within BEAST, used to compute the likelihood of data on a proposed tree, are composed of two main parts: equilibrium frequencies and substitution model.  BEASTling allows users to choose equilibrium frequencies to be uniform, which means that all values have equal equilibrium probabilities. However, by default the inference procedure assumes empirical frequencies, where each value's equilibrium probability is proportional to its frequency in the dataset.  For the substitution model which specifies the relative rates of transition between two unequal states, several different models are available.

A number of substitution models suitable for use in linguistic analyses are available for BEASTling analyses.  The Lewis Mk model\cite{Lewis2001} is a generalised Jukes-Cantor model suitable for discrete features with an arbitrary fixed number of permitted states.  Transitions are permitted from any state to any other, and all transitions are equally likely.  This provides a simple and sensible default model for many types of linguistic data, including typological data from sources such as WALS\cite{Dryer2013} and SAILS\cite{Muysken2014}.  The Bayesian Stochastic Variable Selection (BSVS) model is a refinement of the Mk model in which some transitions may be more likely than others, and further some transitions are explicitly disallowed.  Precisely which transitions are disallowed is inferred during the MCMC analysis.  This model permits searching for directional preferences in the evolution of linguistic features.  The binary Covarion model\cite{Penny2001} is defined for datasets where each feature has two permitted values, 0 and 1.  The model permits a feature to transition between latent ``fast'' and ``slow'' states, which influence the rate at which transitions between 0 and 1 are permitted (transitions in either direction are equally probable).  During the analysis, BEAST will estimate the rate at which features switch between the fast and slow states, and the difference in speed between the two states.  This model has previously been applied to binary-encoded cognate data \cite{Gray2009,Bouckaert2012}.  When this model is used, ascertainment correction for the absence of constant features in the dataset is enabled (unless the user has specifically requested that constant features be retained).  In addition to these standard choices, if a user develops a new substitution model for BEAST, they can also utilise this in a BEASTling analysis by writing a simple Python class.

It is possible to specify multiple datasets in a single BEASTling analysis, and each dataset may be assigned a different substitution model.  This allows combining cognate data and structural data in a single analysis, with appropriate models used for each (binary Covarion and Mk or BSVS, respectively).  Alternatively, this approach allows partitioning one dataset into multiple parts and treating them differently, e.g. using the BSVS model for word order features in a structural dataset and the Mk model for phonological features.  If the languages present in multiple datasets do not match exactly, BEASTling can produce analyses using either the intersection or the union of the languages involved.  If the union is specified, languages will be treated as having missing data for all features defined in datasets from which they are absent.

In addition to substitution models for fitting trees to linguistic data, BEASTling also supports a spherical diffusion model\cite{Bouckaert2016} for fitting trees to location data.  This models can be used to infer posterior regions for the homelands of language families, similar to a previous phylogeographic analysis of the Indo-European language family\cite{Bouckaert2012}.

\subsection{Data formats}

BEASTling accepts linguistic data in comma separated value (CSV) or tab separated value (TSV) file formats, which represent a ``lowest common denominator'' of formats, as opposed to specialist formats like NEXUS.  Two CSV formats are supported.  One is a simple ``native'' format designed for BEASTling, in which each row corresponds to one language, and contains the datapoints for every linguistic feature in the dataset (each column corresponds to one feature).  The other is the Cross-Linguistic Data Format\cite{Forkel2016}, a format standardizing exchange of data within the Cross-Linguistic Linked Data project and related database projects (see \url{http://cldf.clld.org/} for more on CLDF).  This makes it extremely easy to specify analyses using data from existing databases which use the CLDF format, such as WALS, SAILS and eWAVE\cite{Kortmann2013}.

BEASTling aims to provide flexibility in the encoding of data, to minimise the amount of work required to use existing data files.  As long as the data file conforms to one of the two supported CSV file formats, names for languages and features can be arbitrary strings (though each language and feature must have a unique name).  Feature values similarly can be arbitrary strings, with distinct strings being treated as distinct values.  Thus, BEASTling will readily accept data files where different feature values are coded numerically ($0, 1, 2,\ldots$ or $1, 2, 3,\ldots$), alphabetically (A, B, C, \ldots) or in any consistent freely written form (SOV, SVO, VSO, VOS, OVS, OSV for word order, for example).  Missing data points can be represented by question marks.  If the binary Covarion substitution model is selected, data is automatically recoded into the required binary format.  This approach permits BEASTling to share a single substitution rate across the newly created features, and also means the same data file can be used for binary or non-binary substitution models, rather than the user being forced to maintain two copies of the same data for use with different models.  However, if BEASTling is provided with binary data (be it pre-binarised cognate data or typological data with two permissible values), it will recognise this and avoid a secondary binarisation.

\subsection{Glottolog integration}

BEASTling features integration with the Glottolog language catalog\cite{Hammarstroem2016}, which ``aims to provide a comprehensive list of languoids (families, languages, dialects)''.  The 7,748 spoken languages in Glottolog are classified in a principled manner into a collection of phylogenetic trees comprising some 241 families with more than one member.  All clades in the Glottolog classification are provided with names (e.g. Standard English belongs to the Macro-English clade, which is a subclade of Mercian, which is a subclade of Anglian, and so on, up through West Germanic, Northwest Germanic and Germanic to Indo-European) as well as with alphanumeric identifiers called ``glottocodes'' (e.g. Standard English is <tt>stan1293</tt>).  Because the Glottolog classification is made available under a Creative Commons Attribution-ShareAlike 3.0 license, it is possible to include a machine-readable copy of the classification in BEASTling.  If users ensure that languages in their datafiles are referred to either by their three letter ISO-639-3 code or by the four character and four digit glottocode, then BEASTling is aware of the Glottolog classification of the languages in the dataset, and this facilitates several useful features.

Most importantly BEASTling is able to impose monophyly constraints on its analyses which enforce consistency with the Glottolog classification.  This means that, for example, in a BEASTling analysis involving the Indo-European languages, trees in which the Germanic, Romance, Slavic languages etc. are not appropriately organised into distinct clades will be assigned a prior probability of zero.  This ensures that whatever posterior results are produced are guaranteed to be consistent with a widely recognised and respected pre-existing expert classification.

BEASTling's Glottolog integration also makes it very convenient to add important details to configurations.  Any clade in the Glottolog classification can be referred to by name or glottocode for the purposes of imposing calibration dates or selecting subsets of data files.  This makes it quick and convenient to add important pre-existing linguistic knowledge to analyses.  For example, BEASTling recognises ``Imperial Latin'' as a particular node within the Indo-European clade, and we can refer to it by name to use the known age of the Roman empire to provide a calibration date range, rather than having to manually list all of the languages in our analysis which are descended from Imperial Latin.

Glottolog clade names can also be used to specify a subset of languages in a dataset to use for an analysis, enabling the user to easily extract, say, only the Indo-European and Uralic languages from a global database.  This removes the need to maintain multiple copies of the same data for use with modelling different sets of language families.

Finally, Glottolog also provides latitude and longitude location data from the vast majority of its languages.  BEASTling automatically uses this data to specify leaf node locations for phylogeographic analyses, removing the need for the daunting task of researching and manually enter locations for each language in a large analysis.  Alternative location data for families can be easily loaded from a CSV file, overriding the Glottolog data if required.

These features demonstrate the importance and the power of making broadly useful linguistic data available in machine-readable formats under permissive copyright licenses and of referring to languages using standardised, unique identifiers.

Glottolog is an ongoing effort, and names and classifications are subject to change between releases.  Each release of BEASTling will be packaged with the latest Glottolog release available at the time (currently 2.7).  However, any previous Glottolog release back to 2.4 can be specified in a BEASTling configuration file, and BEASTling will download the appropriate data from \texttt{glottolog.org}.  This allows published analyses to be accurately replicated and modified by future releases of BEASTling, even after new Glottolog releases have been made.

\subsection{Advanced features}

In addition to functioning as a command-line program for transforming BEASTling configuration files into BEAST analysis specifications, BEASTling can also be used as a library from within Python scripts.  When used in this fashion, it is possible to generate BEAST XML files without first creating a BEASTling configuration file.  Instead, the high-level analysis parameters are specified as attributes of a Python object.  This is convenient for programmatically generating large numbers of BEAST XML files in which either the same model specification is used for several different datasets, or a range of slightly different model specifications are used for one dataset.  This is particularly useful for large simulation studies.

Similar power for quickly generating a large number of BEAST XML files can be achieved using features of Unix shells, such as \texttt{bash}.  BEASTling can read model specification details from a BEASTling configuration file as per normal, but read the linguistic data from \texttt{stdin}.  In this way, a shell \texttt{for} loop can be used to feed several data files to separate invocations of the \texttt{beastling} command which all reference the same configuration file.

\section{Example analyses}

To illustrate the sorts of analyses BEASTling is designed to facilitate, we present the results of two example analyses.  Our intent is to concisely demonstrate the various abilities of the software, and these analyses should not be construed as serious attempts at historical linguistic scholarship.  The BEASTling configuration files for both analyses are included in the Supplementary Material.

\subsection{Estimating Indo-European family tree from cognate data}

Our first example is an inference of a phylogenetic tree for the Indo-European language family, using cognate data and the binary Covarion model.  The dataset\cite{List2014a} (prepared by List\cite{List2014} using material from the ``Tower of Babel'' project\cite{Starostin2008}) is a comparatively small one which contains 19 languages and 110 features, each of which corresponds to a word meaning.  The datapoints are cognate class assignments, coded as integers.  That is, two languages have the same integer for a given meaning if their words for that meaning are cognate.  Known borrowings are indicated by negative values, i.e. a datapoint of -4 indicates that a language has borrowed a meaning from cognate class 4.  Before running the analysis, we replace all known borrowings with question marks, so that they are treated by BEAST as missing data.  Seven meanings in the dataset are automatically removed by BEASTling because they are constant for the 19 languages included, and thus cannot provide information about the tree topology\footnote{The removed meaning classes are \emph{claw, name, new, salt, two, what} and \emph{who}}.  Because the binary Covarion model was specified, BEASTling automatically reformats the cognate data for the 103 remaining meanings into binary form, resulting in 645 binary features.  Because the languages in the datafile are identified by English names (``Dutch'', ``Swedish'', ``English'', etc.) and not ISO codes or Glottocodes, BEASTling cannot automatically impose monophyly constraints, so this feature is disabled.  No calibration dates are provided, and rate variation across features is enabled.

The maximum clade credibility tree produced by this analysis is shown in Figure \ref{fig:ie}.  Note that despite the lack of monophyly constraints, the tree is in fairly good agreement with conventional wisdom on Indo-European history.  The Slavic, Germanic and Romance sub-families are all correctly positioned in their own clades with high posterior support, and the Slavic family branches before the other two.  The order in which Armenian, Greek and Hindi branch differs from previous analysis\cite{Gray2003,Bouckaert2012}, but this can be readily attributed to the small number of languages in the dataset (note that the clade combining Armenian and Greek has the lowest posterior support).

\begin{figure}[t]
	\begin{center}
	\includegraphics[width=0.45\textwidth]{examples/indoeuropean/mcct.pdf}
	\end{center}
	\caption{Maximum clade credibility tree for the Indo-European languages in our first example analysis.  The values at branches represent posterior probabilities for the corresponding clades.}
\label{fig:ie}
\end{figure}

In addition to a posterior sample of trees, the analysis logs posterior distributions over the relative substitution rate parameters for the 103 features.  A modest amount of rate variation is inferred, with the fastest feature under going change 4 times faster than the slowest feature.  Table \ref{tab:ie} shows the features with the ten highest and ten lowest rates.  
\begin{table}[t]
	\begin{center}
		\input{examples/indoeuropean/table.tex}
	\end{center}
	\caption{Relative substitution rates of the ten slowest and fastest changing meaning classes in our example analysis of Indo-European cognate data.}
\label{tab:ie}
\end{table}

The rates obtained from the analysis allow us to order these lexical features by stability.
Previously, very similar word lists ordered by purported stability have been published by Swadesh\cite{Swadesh1955}, Starostin\cite{Starostin2007} and Pagel et. al.\cite{Pagel2007}.
Using electronic versions of these rankings available from the Concepticon project\cite{List2015} we calculate the Spearman rank correlation coefficients between them and the ranking derived from our posterior mean substitution rates. The coefficients against the Swadesh and Starostin rankings are 0.45 and 0.43, respectively, and a slightly higher 0.56 against the mean of these two rankings (it is worth noting that both these authors based their conclusions on larger data sets than ours).  Against Pagel et. al.'s results, which are also based on Bayesian analysis of only Indo-European languages, we get our highest correlation, of 0.65.
% How good are these numbers? How well-correlated are these lists to each other?

\subsection{Fitting substitution rates to WALS features using a fixed Austronesian tree}

Our second example is an illustration of BEASTling's ability to accept a user-specified starting tree, and to sample only model parameters during the MCMC run, leaving tree topology and branch lengths fixed.  We use the maximum clade credibility tree from a 2009 Bayesian investigation of the Austronesian language family phylogeny by Gray et. al.\cite{Gray2009} as the fixed tree, and typological features from WALS \cite{Dryer2013} as the data.

We label the leaves of the reference tree with ISO codes, and BEASTling automatically prunes the tree to include only those languages whose ISO codes are present in the WALS database.  We configure BEASTling to exclude any features which have known values for less than 25\% of languages.  We also manually exclude 3 WALS features (IDs 95A, 96A and 97A) which are not features in their own right, but instead encode the relationship between other features.  The final analysis involves 169 Austronesian languages and 25 WALS features (see Supplementary Material for a full discussion of the languages and features involved).  A Lewis Mk model is specified for the data, with a log-Normal relaxed clock model and rate variation across features enabled.  The inferred per-feature substitution rates are the subject of interest.  Since the tree is fixed to a known value, BEASTling automatically disables tree logging to save disk space.

The inferred rates of change of these typological features show a much wider variation than the lexical rates of change in the Indo-European example above.  The fastest changing feature has a rate over 11 times higher than the slowest changing feature.  Table \ref{tab:austro} shows the 10 slowest and fastest changing features.  Many of the slowest features are categorised by WALS as word order features, consistent with conventional wisdom that these are some of the most stable typological features.

\begin{table*}[ht]
	\begin{center}
		\input{examples/austronesian/table.tex}
	\end{center}
	\caption{Relative substitution rates of the ten slowest and fastest changing features in our example analysis of Austronesian typological data.}
\label{tab:austro}
\end{table*}

\section{Availability and Future Directions}

BEASTling is an open source project and full source code is available in a version control repository hosted by GitHub at \texttt{https://github.com/lmaurits/BEASTling}, under the terms of a 2-clause BSD license.  BEASTling is also hosted at the Python Package Index (PyPI) and thus may be easily installed using standard Python packaging tools such as \texttt{easy\char`_install} or \texttt{pip}.  Searchable documentation, including a tutorial, is hosted by Read The Docs at \texttt{https://beastling.readthedocs.org}.

The authors intend to continually update BEASTling to support any new linguistically-relevant BEAST packages which may appear, and to keep model specifications in line with emerging consensuses on best practice.  Contributions from the historical computational linguistics research community are welcomed.

\section{Acknowledgemnts}

We are grateful to Simon Greenhill and Mattis List for comments and suggestions on draft versions of this paper.  Thanks to Michael Dunn for testing early versions of BEASTling and submitting bug reports and feature requests; to Simon Greenhill for providing the Austronesian reference tree and associated material; to Mattis List for providing the Indo-European data set and advice on its analysis; and special thanks to Remco Bouckaert for BEAST advice and insight.

This work was supported by a Royal Society of New Zealand Marsden Fund Grant (\#13-UOA-121).

\bibliography{bibliography}
\bibliographystyle{plain}

%\printbibliography
\end{document}

%%% Local Variables: 
%%% coding: utf-8
%%% mode: latex
%%% TeX-engine: xetex
%%% End: 
